src/Agda2Hs/Compile/Function.hs

```diff
@@ -100,19 +100,32 @@ compileFun, compileFun' :: Bool -> Definition -> C [Hs.Decl ()]
 compileFun withSig def@Defn{..} = withFunctionLocals defName $ compileFun' withSig def
 -- inherit existing (instantiated) locals
 compileFun' withSig def@(Defn {..}) = do
-  reportSDoc "agda2hs.compile" 6 $ "compiling function: " <+> prettyTCM defName
+  reportSDoc "agda2hs.compile" 6 $ "Compiling function: " <+> prettyTCM defName
   when withSig $ whenJustM (liftTCM $ isDatatypeModule $ qnameModule defName) $ \_ ->
     genericDocError =<< text "not supported by agda2hs: functions inside a record module"
   let keepClause = maybe False keepArg . clauseType
   withCurrentModule m $ setCurrentRange (nameBindingSite n) $ do
-    ifM (endsInSort defType) (ensureNoLocals err >> compileTypeDef x def) $ do
+    ifM (endsInSort defType)
+        -- if the function type ends in Sort, it's a type alias!
+        (ensureNoLocals err >> compileTypeDef x def) 
+        -- otherwise, we have to compile clauses.
+        $ do
       when withSig $ checkValidFunName x
       compileTopLevelType withSig defType $ \ty -> do
-        -- Instantiate the clauses to the current module parameters
+        let filtered = filter keepClause funClauses
+        weAreOnTop <- isJust <$> liftTCM  (currentModule >>= isTopLevelModule)
         pars <- getContextArgs
-        reportSDoc "agda2hs.compile" 10 $ "applying clauses to parameters: " <+> prettyTCM pars
-        let clauses = filter keepClause funClauses `apply` pars
+        -- We only instantiate the clauses to the current module parameters
+        -- if the current module isn't the toplevel module
+        unless weAreOnTop $
+          reportSDoc "agda2hs.compile.type" 6 $ "Applying module parameters to clauses: " <+> prettyTCM pars
+        let clauses = if weAreOnTop then filtered else filtered `apply` pars
```

Comment:
Could someone confirm that this is sufficient logic? It would appear top-level parameters are never part of a function clauses.